### ISSUE TITLE
New rule proposal

### ISSUE THREAD

**Issue by hornta** [yannickcr/eslint-plugin-react#2281]
This rule would report on unreachable code paths in React components.

Example: this.state.bar will never be truthy because there is no `setState({ bar: ... });` and will make the code inside the condition unreachable.

Example of **incorrect** code
```javascript
class Foo extends React.Component {
  constructor() {
    this.state = {
      bar: false
    };
  }

  render() {
    if(this.state.bar) {
      return 'bar'; // this is unreachable
    }
    return 'foo';
  }
}
```

```javascript
function Foo() {
  const [bar, setBar] = useState(false);

  if(bar) {
    return 'bar'; // this is unreachable
  }

  return 'foo';
}
```

```javascript
function Foo() {
  const [bar, setBar] = useState(false);

  setBar(null); // setting to null still makes the code unreachable

  if(bar) {
    return 'bar'; // this is unreachable
  }

  return 'foo';
}
```

Example of **correct** code
```javascript
class Foo extends React.Component {
  constructor() {
    this.state = {
      bar: false
    };
  }

  componentDidMount() {
    this.setState({ bar: true }); // setting bar to a known non-falsy value
  }

  render() {
    if(this.state.bar) {
      return 'bar'; 
    }
    return 'foo';
  }
}
```

```javascript
function Foo() {
  const [bar, setBar] = useState(false);

  setBar(returnWhat()); // because returnWhat can return a truthy value

  if(bar) {
    return 'bar'; // this may be reachable, we can't know
  }

  return 'foo';
}
```

**Comment by ljharb** [comment]
That sounds very difficult to code reliably; since `this.setState` can be passed any object literal, which wouldn't be statically analyzeable - iow, any setState call that didn't have an object literal in it, or that had spread syntax in it, would make the rule useless for the file.
